text/0000-bee-model-crate/0000-bee-model-crate.md

```diff
@@ -0,0 +1,279 @@
+
+Each transaction is identified by its hash, the transaction hash. Transactions should be final and therefore not modifiable after construction.
+The same applies to bundles. Bundles are final, and therefore not modifiable after construction.
+Having bundles as well as transactions final helps achieve correctness among the data objects.
+
+Transactions can be built from Transaction Builders.
+Bundles can be built from Bundle Builders. Both builder objects can be manipulated as desired.
+It should be noted, Bundle Builders don't process Transactions (as transactions are final), instead, they process Transaction Builders.
+
+## Exposed Interface
+
+### Transaction
+
+ A transaction consists of following fields:
+
+```rust
+struct Transaction {
+
+    transaction_hash: String,
+    signature_fragments: String,
+    address: String,
+    value: i64,
+    obsolete_tag: String,
+    timestamp: i64,
+    current_index: usize,
+    last_index: usize,
+    bundle_hash: String,
+    trunk: String,
+    branch: String,
+    nonce: String,
+    tag: String,
+    attachment_timestamp: i64,
+    attachment_timestamp_lower_bound: i64,
+    attachment_timestamp_upper_bound: i64,
+
+}
+```
+
+As mentioned, Transactions are final. Transaction fields therefore should be only accessible by getter functions.
+Besides that, Transactions can be only built by Transaction Builders, or by the bytes of a received transaction.
+
+```rust
+impl Transaction {
+
+    pub fn from_transaction_builder(transaction_builder: &TransactionBuilder) -> Result<Transaction, TransactionBuilderValidationError> {
+                
+        match TransactionBuilderValidator::validate(&transaction_builder) {
+            Ok(()) => { ; }
+            Err(e) => { return Err(e) }
+        }
+
+        let (transaction_hash, nonce, timestamp) = Pow::compute(&transaction_builder);
+
+        Ok(Transaction {
+            transaction_hash,
+            signature_fragments: transaction_builder.signature_fragments.clone(),
+            address: transaction_builder.address.clone(),
+            value: transaction_builder.value.clone(),
+            obsolete_tag: transaction_builder.obsolete_tag.clone(),
+            timestamp,
+            current_index: transaction_builder.current_index.clone(),
+            last_index: transaction_builder.last_index.clone(),
+            bundle_hash: transaction_builder.bundle_hash.clone(),
+            trunk: transaction_builder.trunk.clone(),
+            branch: transaction_builder.branch.clone(),
+            nonce,
+            tag: transaction_builder.tag.clone(),
+            attachment_timestamp: transaction_builder.attachment_timestamp.clone(),
+            attachment_timestamp_lower_bound: transaction_builder.attachment_timestamp_lower_bound.clone(),
+            attachment_timestamp_upper_bound: transaction_builder.attachment_timestamp_upper_bound.clone(),
+        })
```

Comment:
I think this way of creating the transaction is dangerous. I think a TransactionBuilder should be destroyed having a method `build(self) -> Transaction` moving the internal state (the transaction to be built) out. I seems like what you want to have is a `TransactionFactory`, that has some template and creates clones off of it. But since transactions are completely unique instances where nothing can be really shared across them, the factory pattern doesn't seem to be a good choice here.

Reply:
Thanks for the comment. Agree, moving ownership and destroy it afterwards seems to be the better choice.